### DISCUSSION TITLE
Why is tRPC not listening allowing me to listen to subscriptions more than 1 time -React Native.

### DISCUSSION THREAD

**Discussion by CrispenGari** [trpc/trpc#4945] [Q&A] [answered]
I have a subscription server with trpc that is working. But when i try to subscribe for the top component the subscription works but in if i try to run the subscription on the sub-component or running the same subscription on which has already been subscribed to I'm not getting the data back.

I'm subscribing to like this in my `App.ts`:
```ts
  trpc.notification.onDelete.useSubscription(
    { uid: me?.id || "" },
    {
      onData: async (_data) => {
        await refetch();
      },
    }
  );
```
> When new data is in this subscription works. However in the other component that I'm using this subscription `Notification.tsx` this is not returning data. 

```tsx
  trpc.notification.onDelete.useSubscription(
    { uid: me?.id || "" },
    {
      onData: async (_data) => {
        await refetch();
      },
    }
  );
```
I believe the problem is with `trpc`. Can i listen to the same subscription twice,?

**Comment by dhmoon91** [comment]
Same I'm not getting anything using subscription. 

Works fine on web for the same methods and usage tho.

**Comment by KATT** [comment]
You should be able to subscribe many times and each subscription should create a new request (we don't dedupe subscriptions)

We can't help you without a reproduction

**Reply by KATT** [reply]
Maybe there's some sort of max concurrent connection limit in RN? I'm not sure

**Comment by dhmoon91** [comment]
I think I got it working - didn't test on production yet but working fine locally.
Looks like react-native, by default, doesn't support Eventstream that needs to be configured for any SSE to work. 

You have to install custom module (I think they polyfill it so react native can understand), and set the connection manually.

I installed `react-native-sse` and configured a custom hook to listen to the event directly instead of using trpc client. 

here's the hook code: 

```
import EventSource from 'react-native-sse';

export function useTRPCSubscription(topic: string | number) {
  const [data, setData] = useState<any | null>(null);


  // UPDATE url per your need.
  useEffect(() => {
    const test = `${url}/chat.onMessage?input=${encodeURIComponent(JSON.stringify({ roomId: topic }))}&connectionParams=${encodeURIComponent(JSON.stringify({ token: trpcHeaderToken }))}`;

    const eventSource = new EventSource(`${test}`);

    eventSource.addEventListener('message', (event: any) => {
      const message = JSON.parse(event.data);
      setData(message);
    });

    eventSource.addEventListener('open', () => {
      console.log('SSE connection opened');
    });
    eventSource.addEventListener('error', (err) => {
      console.error('SSE error:', err);
      eventSource.close();
    });

    return () => {
      eventSource.close();
    };
  }, [topic]);

  return data;
}

```

Use it on actual page; 
```
const ChatScreen: FC = () => {
  const [messages, setMessages] = useState<Message[]>([]);
  const roomId = 2;
  ... 
  const newMessages = useTRPCSubscription(roomId);

useEffect(() => {
    if (newMessages) {
      setMessages((prev) => [...prev, newMessages as Message]);
      setTimeout(() => {
        scrollViewRef.current?.scrollToEnd({ animated: true });
      }, 100);
    }
  }, [newMessages]);
  ...
}
```

**Reply by BeBoRE** [reply]
https://trpc.io/docs/client/links/httpSubscriptionLink#compatibility-react-native maybe worth taking a look at the docs for rn, maybe that works better. Would love to know what you think.

**Reply by dhmoon91** [reply]
oh nice - I was originally looking for similar solution. 
Didn't know TRPC had doc on it under `httpsubscriptionlink` - just tried it and seems to work fine Thanks @BeBoRE !

**Reply by dhmoon91** [reply]
oh actually - I rebuilt the app and I'm seeing same `'EventSource' doesn't exist error.`
Im just gonna stick to original hacky solution I had for now.

**Reply by BeBoRE** [reply, marked as answer]
The EventSource needs to be ponyfilled as seen in https://trpc.io/docs/client/links/httpSubscriptionLink#custom-headers-through-ponyfill with the polyfill suggested in the rn section. Just realizing now that's not shown in that section.

**Reply by dhmoon91** [reply]
yer thanks!

**Reply by bikoevD** [reply]
> The EventSource needs to be ponyfilled as seen in https://trpc.io/docs/client/links/httpSubscriptionLink#custom-headers-through-ponyfill with the polyfill suggested in the rn section. Just realizing now that's not shown in that section.

following these instructions - https://trpc.io/docs/client/links/httpSubscriptionLink#compatibility-react-native:
```
globalThis.ReadableStream = globalThis.ReadableStream || ReadableStream;
globalThis.TransformStream = globalThis.TransformStream || TransformStream;
globalThis.EventSource = RNEventSource
```
i got ts error for EventSource and had to add static properties:
```
RNEventSource.CONNECTING = 0;
RNEventSource.OPEN = 1;
RNEventSource.CLOSED = 2;

globalThis.EventSource = RNEventSource as unknown as typeof EventSource;
```
